frontend/catalyst/jax_extras/lowering.py

```diff
@@ -165,3 +166,54 @@ def custom_lower_jaxpr_to_module(
                 worklist += [*op.body.operations]
 
     return ctx.module, ctx.context
+
+
+with ir.Context() as ctx:
+    with ir.Location.unknown(ctx) as loc:
```

Comment:
You can put these two lines in the same line.

```suggestion
with ir.Context() as ctx, ir.Location.unknown(ctx) as loc:
```

I've encountered issues in the past where types from different contexts are not compared correctly. E.g., i64 from ctx1 was different form i64 from ctx2. This code is not used anywhere because none of our passes have options in them at the moment. I see the tests, but I wonder whether an integration test would be more useful.

Also, I think this method would benefit from pattern matching. Look into the `match` and `case` statements.

Reply:
> Also, I think this method would benefit from pattern matching. Look into the `match` and `case` statements.

matching against the type looks weird, you'd have to "call" the types
```python
t = 12.0
match t:
  case int():
    print("int")
  case float():
    print("float")
```

so I went with the if statements :sweat_smile:

Reply:
> I've encountered issues in the past where types from different contexts are not compared correctly. E.g., i64 from ctx1 was different form i64 from ctx2.

But we are not comparing types here right? 

And if we want to be safe, I can pass the same `Context()` instance into the recursive calls. Or is it done already since the recursive calls are all under the same `with` block?

Reply:
Just a note, recursive calls for a function under the same `with` do not create new `with` context objects, so we are not actually creating new jax `ir.Context()` instances here
```python
class Thing:
   def __init__(self, thing):
      self.thing = thing
   def __enter__(self):
      return self.thing
   def __exit__(self, _, __, ___):
      return

with Thing("apple") as apple:
   def foo(i):
      print(id(apple))
      if i==0:
         return 0
      foo(i-1)

foo(3)
```

```
130706228697584
130706228697584
130706228697584
130706228697584
```

Reply:
Yes, I agree. But I am still worried about using a new Context as opposed to the one provided by JAX. Maybe I am wrong though. Here is the ticket that has a comment regarding precomputing types and using them in the JAX modules: https://github.com/PennyLaneAI/catalyst/pull/610#discussion_r1556672108

Reply:
Essentially, if we remove the `with Context` and place it in the tests, it would allow for the context used by JAX to be used normally (I think?) Which context do you think would be used?

Reply:
I think you don't need the context not the Location unknown if it already is under the context and location. So I think you can just remove the contextmanagers and keep the definition, but in the tests you have the context managers.

No need to pass the context directly. I believe the context argument is optional when under the context manager.

Reply:
See for example: frontend/catalyst/utils/types.py and frontend/test/pytest/test_types.py